### DISCUSSION TITLE
Local development flow recommendations

### DISCUSSION THREAD

**Discussion by milesingrams** [supabase/supabase#3277] [Questions] [answered]
I'm just starting using Supabase and I would like advice on what best practices for development flows are. Since I'm in early stage development on my project I'm imagining there will be a lot of table changes, grant changes, RPC changes, policy changes, index changes etc happening frequently. I'm trying to figure out the best way to do things.

1. How might I best manage these changes cleanly in local development without doing migrations (Since I don't care about local data. Happy to just drop everything and rerun the DB setup scripts. Is there an easy way to fully reset the local Supabase and delete all data?)

2. Once I'm ready to push the changes to a staging project hosted on Supabase, what suggestions and best practices are there to take the local DB and migrate all the differences to the hosted project. I'm assuming this process would be identitcal to the one to migrate to production.

Any additional tips/thoughts would be majorly appreciated!
Really excited to see the future of Supabase. Love the power of postgres combined with the easy of use and approachability of Firebase.

**Comment by soedirgo** [comment, marked as answer]
Hi Miles!

Local development and schema evolution is exactly the purpose of the [CLI](https://github.com/supabase/cli/tree/new), but it's still at an experimental stage. Being able to change schemas e.g. with an SQL IDE of your choice and then using a [diff tool](https://supabase.io/new/blog/2021/03/31/supabase-cli#diffs) to generate migrations is the experience we're aiming for.

The local Supabase Docker Compose setup is supposed to persist data as opposed to the local dev setup with the CLI which is transient, so they're likely to diverge. But for migrating the diff, you may be able to use the diff tool on its own and cherry-pick the resulting DDL.

FTR we're also thinking of a Git-like workflow where you can have multiple transient databases (for staging or feature branches) alongside your main (prod) database in a Supabase project. But again, this is still in the works.

I think creating a great local dev experience is tricky to get right, so let me know if you have further ideas/questions!

**Reply by milesingrams** [reply]
Thanks for your insight!
The Git-like workflow sounds ideal and very excited for that to be developed/available.
I guess I'm still confused as to the difference between the Docker Compose setup and the CLI. Currently I am using the CLI (`supabase start/stop`) and the schemas are retained between stopping and starting the container.
Since I have no data I care to persist between local dev runs and would rather just rebuild the db from a `setup.sql` file with some seed data every time (much like the slack clone example), would the best practice be to drop the `public` schema and rerun the setup file each time? Then once I have local in a state that I like I can use a diff tool to create a migration for staging and prod?

**Reply by milesingrams** [reply]
I guess while a lot of people want the GUI experience I'm going more for editing and maintaining a single or set of SQL files that describe the current state of the schemas/policies/grants/rpcs etc. That way I can version control my 'backend' and know that it exactly matches the file stored in my repository. Maybe this is an antipattern? Advice would be greatly appreciated.

**Reply by soedirgo** [reply]
> the schemas are retained between stopping and starting the container

Yeah, this was how it's done in the old Node.js CLI (which will be EOL'd), but for the new CLI the schema is recreated on start. If you'd like a sneak peek to the new CLI, you can try out a little [tour](https://github.com/supabase/cli/tree/new/examples/tour) I made - would appreciate the feedback :)

> I'm going more for editing and maintaining a single or set of SQL files that describe the current state of the schemas/policies/grants/rpcs etc

Yup, we do this by having a declarative "structured dump" that gets VC'd so you can see how your schema evolves by going through your Git history. (Again, check out the tour to see what it's like!) But this structured dump is write-only - we found that making a declarative state the source of truth not flexible enough to handle trickier migrations, like renaming columns or data migration.

**Reply by Vinzent03** [reply]
@soedirgo 
First, thank you for your work at the cli. I'm currently thinking about my setup. For production, I'm using the hosted version on supabase.io, but of course I don't want to develop on production, so after reading some documentation I ended at the [cli](https://github.com/supabase/cli). Am I right that there is no file containing all sql needed for my setup? Instead, it is split over every single migrations file? So I have to keep them all? 
Now I want to do some additional testing of my app with GitHub actions. What is the best approach for that? I guess it's [self-hosting](https://supabase.io/docs/guides/hosting/docker). But how do I get my database structure to that self-hosting setup?

**Reply by EskelCz** [reply]
@soedirgo But in this setup how can we collaborate on database code with an AI (copilot/cody/devgpt) without a single source of truth in a text file? :-/
